infra/src/main/kotlin/fi/fta/geoviite/infra/tracklayout/LayoutAlignmentService.kt

```diff
@@ -61,6 +78,57 @@ class LayoutAlignmentService(private val dao: LayoutAlignmentDao) {
             }
         }
     }
+
+    fun getOverlappingPlanHeaders(
+        alignmentVersion: RowVersion<LayoutAlignment>,
+        geocodingContextCacheKey: GeocodingContextCacheKey,
+        startKmNumber: KmNumber?,
+        endKmNumber: KmNumber?,
+    ): List<GeometryPlanHeader> {
+        val croppedAlignment = cropAlignment(alignmentVersion, geocodingContextCacheKey, startKmNumber, endKmNumber)
+        val simplified = simplify(croppedAlignment, ALIGNMENT_POLYGON_SIMPLIFICATION_RESOLUTION, null, true)
+
+        val polygon = bufferedPolygonForLineStringPoints(simplified, ALIGNMENT_POLYGON_BUFFER, LAYOUT_SRID)
+        val plans = geometryDao.fetchIntersectingPlans(polygon, LAYOUT_SRID)
+        return geometryDao.getPlanHeaders(plans)
+    }
+
+    private fun cropAlignment(
+        alignmentVersion: RowVersion<LayoutAlignment>,
+        geocodingContextCacheKey: GeocodingContextCacheKey,
+        startKmNumber: KmNumber?,
+        endKmNumber: KmNumber?,
+    ): IAlignment {
+        val alignment = dao.fetch(alignmentVersion)
+        if (startKmNumber == null && endKmNumber == null) return alignment
+
+        val geocodingContext = geocodingService.getGeocodingContext(geocodingContextCacheKey)
+        val startM =
+            requireNotNull(
+                startKmNumber?.let {
+                    geocodingContext?.referencePoints?.find { it.kmNumber == startKmNumber }?.distance
+                } ?: alignment.start?.m
+            )
+        val endM =
+            requireNotNull(
+                endKmNumber?.let { geocodingContext?.referencePoints?.find { it.kmNumber > endKmNumber }?.distance }
+                    ?: alignment.end?.m
+            )
+
+        val startSegment =
+            alignment.segments
+                .find { s -> startM in s.startM..s.endM }
+                ?.let { s -> if (startM < s.endM) s.slice(Range(startM, s.endM)) else null }
+        val endSegment =
+            alignment.segments
+                .find { s -> endM in s.startM..s.endM }
+                ?.let { s -> if (s.startM < endM) s.slice(Range(s.startM, endM)) else null }
+        val midSegments = alignment.segments.filter { s -> s.startM >= startM && s.endM <= endM }
+
+        val segments = (listOfNotNull(startSegment) + midSegments + listOfNotNull(endSegment)).distinctBy { it.id }
```

Comment:
Yksinkertaistus pelaa nyt aika suurelti referencepointeilla. Jos rajaus on annettu, rajataan alignment alkamaan sen referencepointin m-arvosta jolla on `startKmNumber`:in mukainen ratakilometri ja loppumaan `endKmNumber`:ia _seuraavan_ ratakilometrin referencepointille (tällöin koko loppuratakilometri tulee mukaan). Jos arvoja ei ole annettu, defaultataan ne alignmentin alkuun/loppuun.

Segmentit filtteröidään siten, että alun ja lopun segmentit koitetaan leikata jos on tarve. Niiden väliset segmentit otetaan sellaisinaan.

Note to self: Tässä täytyisi ottaa vielä huomioon se, että alku- ja loppupisteet osuvat samalle segmentille

Reply:
Todennäköisesti riittäisi varmaankin joo vain yksi. Se aiheuttaa vähän lisämonimutkaisuutta sen tulosjoukon käsittelyyn (nuo samat slicetystemput pitää tehdä edelleen ekalle ja vikalle, ja niiden väliset segmentit pitää jättää rauhaan) eikä tämä nykytapakaan ole ollut varsinaisesti erityisen hidas, mutta ehkä joku tuollainen kikkailu saattaa kumminkin olla vaivan arvoista 🤔 tutkin asiaa